faiss-search/scann/scann.py

```diff
@@ -0,0 +1,85 @@
+
+from jina.executors.indexers.vector import BaseNumpyIndexer
+
+
+class ScannIndexer(BaseNumpyIndexer):
+    """Scann powered vector indexer
+
+    For more information about the Scann supported parameters, please consult:
+        - https://github.com/google-research/google-research/tree/master/scann
+
+    .. note::
+        Scann package dependency is only required at the query time.
+    """
+
+    def __init__(self, train_filepath: str = None,
+                 num_leaves: int = 2000,
+                 num_leaves_to_search: int = 100,
+                 training_iterations: int = 10,
+                 distance_measure: str = "dot_product",
+                 training_sample_size: int = 250000,
+                 scoring: str = "score_ah",
+                 anisotropic_quantization_threshold: float = 0.2,
+                 dimensions_per_block: int = 2,
+                 reordering_num_neighbors: int = 100,
+                 *args, **kwargs):
+        """
+        Initialize an ScannIndexer
+
+        :param train_filepath: the training data file path.
+        :param num_leaves: The higher num_leaves, the higher-quality the partitioning will be.
+            It should be roughly the square root of the number of datapoints.
+        :param training_iterations: Number of iterations per training. Default is 10
+        :param distance_measure: The distance measurement used betwent the points.
+            It can be dot_product or squared_l2
+        :param num_leaves_to_search: The amount of leaves to search.
+            This should be tuned based on recall target
+        :param training_sample_size: The size of the training sample.
+            Default is 25k
+        :param scoring: It can be score_ah (asymmetric hashing) or score_bf for brute force.
+            For small datasets (less than 20k) brute force is recommended
+        :param anisotropic_quantization_threshold: See https://arxiv.org/abs/1908.10396 for
+            a description of this value
+        :param dimensions_per_block: Recommended for AH is 2
+        :param reordering_num_neighbors: Should be higher than the final number of neighbors
+            If this number is increased, the accuracy will increase but it will impact speed
+        :param args:
+        :param kwargs:
+        """
+        super().__init__(*args, **kwargs)
+        self.train_filepath = train_filepath
+        self.num_leaves = num_leaves
+        self.training_iterations = training_iterations
+        self.distance_measure = distance_measure
+        self.num_leaves_to_search = num_leaves_to_search
+        self.training_sample_size = training_sample_size
+        self.scoring = scoring
+        self.anisotropic_quantization_threshold = anisotropic_quantization_threshold
+        self.dimensions_per_block = dimensions_per_block
+        self.reordering_num_neighbors = reordering_num_neighbors
+
+    def build_advanced_index(self, vecs: 'np.ndarray'):
+        """Load vectors into Scann indexers
```

Comment:
They are using `Command Pattern` (https://sourcemaking.com/design_patterns/command). We should comment that is a lazy evaluation. That `score_ah` and `reorder` are creating just `configuration` to be passed to `create_py_bind` which is the one actually building the object.